docs/core/v3.0.rst

```diff
@@ -0,0 +1,1761 @@
+    is ``false``, then the extension may be ignored and processing may
+    continue. If the extension is not recognized and the value of
+    ``must_understand`` is ``true`` then processing must terminate and
+    an appropriate error raised.
+
+For example, below is an entry point metadata document, specifying that
+JSON is being used for encoding of group and array metadata::
+
+    {
+        "zarr_format": "https://purl.org/zarr/spec/core/3.0",
+        "metadata_encoding": "https://purl.org/zarr/spec/core/3.0",
+        "metadata_key_suffix" : ".json",
+        "extensions": []
+    }
+
+For example, below is an entry point metadata document as above, but also
+specifying that an extension is being used which may be
+ignored if not understood::
+
+    {
+        "zarr_format": "https://purl.org/zarr/spec/core/3.0",
+        "metadata_encoding": "https://purl.org/zarr/spec/core/3.0",
+        "metadata_key_suffix" : ".json",
+        "extensions": [
+            {
+                "extension": "http://example.org/zarr/extension/foo",
+                "must_understand": false,
+                "configuration": {
+                    "foo": "bar"
+                }
+            }
+        ]
+    }
+
+
+.. _array-metadata:
+
+Array metadata
+--------------
+
+Each Zarr array in a hierarchy must have an array metadata
+document. This document must contain a single object with the
+following mandatory names:
+
+``shape``
+^^^^^^^^^
+
+    An array of integers providing the length of each dimension of the
+    Zarr array. For example, a value ``[10, 20]`` indicates a
+    two-dimensional Zarr array, where the first dimension has length
+    10 and the second dimension has length 20.
+
+``data_type``
+^^^^^^^^^^^^^
+
+    The data type of the Zarr array. If the data type is defined in
+    this specification, then the value must be the data type
+    identifier provided as a string. For example, ``"<f8"`` for
+    little-endian 64-bit floating point number.
+
+    The ``data_type`` value is an extension point and may be defined
+    by an extension. If the data type is defined by an
+    extension, then the value must be an object containing the names
+    ``extension``, ``type`` and ``fallback``. The ``extension`` is
+    required and its value must be a URI that identifies the
+    extension and dereferences to a human-readable representation of
+    the specification.  The ``type`` is required and its value is
+    defined by the extension. The ``fallback`` is optional
+    and, if provided, its value must be one of the data type
+    identifiers defined in this specification. If an implementation
+    does not recognise the extension, but a ``fallback`` is present,
+    then the implementation may proceed using the ``fallback`` value
+    as the data type. For fallback types that do not correspond to base
+    known types, extensions can fallback on a raw number of bytes using
+    the raw type (``r*``).
```

Comment:
Probably fallback would not be supported in that case.

I would say though that the current silent fallback behavior described in the spec is not good, because that means the result may differ silently depending on the version of the library you are using.

Instead, I think the behavior should be:

- If the data type is not supported but fallback is specified and supported, and the user has not explicitly opted into using the fallback data type, then the implementation should display an error indicating that a fallback data type is available.

Reply:
> Variable length data types will probably have more than one buffer associated with an array. How could fallback interact with that?

I agree that there wouldn't be a fallback in the case of variable length data types.

> Does the value for fallback have to be defined within the core specification?

I added a change in #167 to relax that.

> I would say though that the current silent fallback behavior described in the spec is not good, because that means the result may differ silently depending on the version of the library you are using.
>
> Instead, I think the behavior should be:
>
> If the data type is not supported but fallback is specified and supported, and the user has not explicitly opted into using the fallback data type, then the implementation should display an error indicating that a fallback data type is available.

Added a note about this in #167.

---------

Let's continue this discussion in #167